lcls-twincat-pmps/PMPS/AncillaryComponents/FB_JsonDocToSafeBP.TcPOU

```diff
@@ -0,0 +1,252 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_JsonDocToSafeBP" Id="{be21f50f-6e98-4ce2-92ee-0a1b1ced09d9}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_JsonDocToSafeBP
+VAR_INPUT
+	bExecute             : BOOL; //Rising Edge
+	jsonDoc 		: SJsonValue;
+	sDeviceName : STRING;
+	//asDeviceStateNames: ARRAY [1.. MOTION_GVL.MAX_STATES] OF STRING;
+END_VAR
+VAR_OUTPUT
+	bHasDevice		: BOOL;
+	bHasAllStates		: BOOL;
+	bHasAllParameters :BOOL := TRUE;
+	bBusy                : BOOL;
+    bError               : BOOL;
+    nErrId               : UDINT;
+	sErrMsg				 :STRING;
+END_VAR
+VAR_IN_OUT
+	//ARRAY [1.. MOTION_GVL.MAX_STATES] ;
+	arrStates: ARRAY[1..MAX_STATES] OF DUT_PositionState;
+	io_fbFFHWO    :    FB_HardwareFFOutput;
+END_VAR
+VAR
+	(* JSON *)
+	fbJson 			: FB_JsonDomParser;
+	jsonProp		: SJsonValue;
+	jsonValue		: SJsonValue;
+	jsonParam		: SJsonValue;
+	jsonnTran      : SJsonValue;
+	jsonnRate      : SJsonValue;
+	Step: INT;
+	index: INT;
+	nStateCount:INT;
+    RisingEdge           : R_TRIG;
+
+	(*Logger*)
+	tNewMessage : R_TRIG;
+    fbLogger : FB_LogMessage := (eSubsystem:=E_SubSystem.MPS, nMinTimeViolationAcceptable:=10);	
+	(*FFO*)
+	FFO    :    FB_FastFault :=(
+        i_Desc := 'Fault occurs when there is an error loading safe beam parameters from json file',
+        i_TypeCode := 16#FF13);
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[(* Retrieve JSON content *)
+RisingEdge(CLK:=bExecute); 
+
+CASE Step OF 
+    0:     (* Idle state *) 
+          IF RisingEdge.Q THEN 
+               bBusy := TRUE; 
+               bError:= FALSE; 
+               nErrId:=0; 
+               Step := 20; 
+			   fbLogger(sMsg:='Start Parsing Json DOC.', eSevr:=TcEventSeverity.Info);
+          END_IF 
+	
+	 20: (*Find the Device Name*)  
+		bHasDevice := fbJson.HasMember(jsonDoc, sDeviceName);
+		IF (bHasDevice) THEN
+			jsonProp := fbJson.FindMember(jsonDoc, sDeviceName);
+			IF jsonProp <> 0 THEN	
+				Step := Step+1;
+				ELSE
+					Step := 900 + Step;
+					nErrId := Step; 
+					bError := TRUE; 
+					sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);
+				END_IF
+		ELSE
+			Step := 900 + Step;
+			nErrId := Step; 
+			bError := TRUE; 
+			sErrMsg:=  CONCAT('Error device name not found in Json File : ',sDeviceName);		
+		END_IF
+		
+	 21: (*Check device States in DOM*)
+	 		FOR Index:=1 TO MAX_STATES BY 1 DO
+				  IF NOT ((arrStates[Index].sPmpsState= '') OR (arrStates[Index].sPmpsState= 'Invalid')) THEN
```

Comment:
These need to be `sName` or else the comparison doesn't work properly. `sPmpsState` will never be `Invalid`, for example.

Reply:
Thank you! fixed in c5646af

Reply:
Sorry to ask this again, What is the difference between sName and sPmpsState? Which one should be in the Json file?

Reply:
`sName` is the human-readable short name such as "OUT" or "YAG"
`sPmpsState` is the lookup key such as "IM1K0-XTES-YAG"

Reply:
So sPmpsState in the value that will be in the json file. So I should revert the changes and in step 21 check that sPmpsState is in the DOM.

Reply:
The code I was running right before Thanksgiving seemed to do everything correctly in the json reading